vike/shared/page-configs/Config.ts

```diff
@@ -78,7 +81,12 @@ type ConfigBuiltIn = {
    *
    *  https://vike.dev/onBeforeRender
    */
-  onBeforeRender?: Function | ImportString | null
+  onBeforeRender?:
+    | ((pageContext: PageContextServer) => OptionalPromise<{ pageContext: Partial<Vike.PageContext> }>)
+    | ((pageContext: PageContextClient) => OptionalPromise<{ pageContext: Partial<Vike.PageContext> }>)
+    | ((pageContext: PageContext) => OptionalPromise<{ pageContext: Partial<Vike.PageContext> }>)
+    | ImportString
+    | null
```

Comment:
@brillout should be better now. Let me know if not clear why I need three signatures. I can illustrate with a TS Playground if needed

Reply:
👍 I think we're on the same page.

I'm seeing that the user *has* to type his `onBeforeRender()` hook because `Config['onBeforeRender']` is a union. In other words, it seems that, if `Config['onBeforeRender']` has only one function type then TypeScript will infer the type of `onBeforeRender()`.

That's quite a bummer as I was hoping that all the user had to do was to set `const onBeforeRender: Config['onBeforeRender'] = ...` and TypeScript would infer the `pageContext` and return type.

I can't think of a solution 🤔.

Anyways, it's better than `Function` as it ensures the user doesn't use wrong types.

Reply:
Yes and there is another reason why the type should be set explicitly on `onBeforeRender()`, which is because of some otherwise-non-strictness in the return types, which I illustrate and explain [here](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgApwOYQMIHtwQAeYyA3gFDJqYQBcyAzmFKBpcgvjMBvUyyDZUwwMABsIAfj7NW5AL7lyYAJ4AHFAHkQAIQgxcUCACUIIACbQAYgFcQCEfmQBeZAAo1NPAWL10Wb0hiAEoXAD4yZE8A-CCwPzgoETgxAB5-HFiiMAjFck4QJmR8PQMjUwtoAEZ6bVLDEzNLKFt7RxAXd1DnCIoqIzAbKA6+qiivLN8ydjHkEXE6ZAAiAFsVABVRCSWAGiVZqgB6Q+R1gAtgBmQAd0MAayuAIwgEOBsGFFwYZBB8AFp+MAHCAIAwrqBkAMhh1VBobmczMg4GCeCA5AdkMckcgYHYHMAnLCUGBcEjfmAEVAcXj2nN1BAZmNHmI4PQlszcI8lozFFRFHkCkUSvoGhVmgAmWq6EXlJrWGkEjquNzBBJJYApdITHw5cLTfoQQbDfVjaKZHX0UazeYSNlrTYLXb7A5YgByuGucwuVxYGDOYFxYjEKhxcGAYgYOy9cBADzmpIpKCIajEQNEkMN0LpGkZVGZrOWHK5PPY-PIQA). Let me know if unclear :) Otherwise I'll now keep extending this PR with other hook types.

Reply:
Yea I've seen that as well. ~~As much as I wish TypeScript would infer the hook type, it doesn't seem like there is a way to do it that preserves correctness.~~ Actually I thought a bit more about this.

I'm thinking: maybe we can remove the union by setting the onBeforeRender's pageContext to `PageContextServer` while enabling users to change the type by using the global interface `Vike.Config`. Thoughts? Let me know if you feel like it's out of your PR's scope.

Reply:
> which I illustrate and explain [here](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgApwOYQMIHtwQAeYyA3gFDJqYQBcyAzmFKBpcgvjMBvUyyDZUwwMABsIAfj7NW5AL7lyYAJ4AHFAHkQAIQgxcUCACUIIACbQAYgFcQCEfmQBeZAAo1NPAWL10Wb0hiAEoXAD4yZE8A-CCwPzgoETgxAB5-HFiiMAjFck4QJmR8PQMjUwtoHThzem1SwxMzSyhbe0cQF3dQ5wiKKiMwGyhO-qooryzfMnZx5BFxOmQAIgBbFQAVUQllgBolOaoAeiPkDYALYAZkAHdDAGtrgCMIBDgbBhRcGGQQfABafjABwgCAMa6gZCDYadVQaW7nMzIODgnggOSHZAnZHIGB2BzAJxwlBgXDIv5gRFQXH4jrzdQQWbjJ5iOD0ZYs3BPZZMxRURR5ApFEr6RoVFoAcVwuFqyHqovKzWstMJnVcbmC9FIExiPni1CSwBS6UmerC8nCMwGECGIyt42imT1WqZwm2SzWm3dewOh2xADlcDd5pdriwMOcwHixGIVLi4MAxAxdiG4CBHvMyZSUEQ1GJgaIoTaYfSNK7kCy2StOdzeewBeQgA).

I'm also seeing that `satisfies` isn't strict as well. Also, even setting the return type isn't strict, which I find quite surprising (see `badAsWell: false`.) I udated your playground to showcase what I mean [here](https://www.typescriptlang.org/play?ssl=32&ssc=1&pln=33&pc=1#code/JYOwLgpgTgZghgYwgAgApwOYQMIHtwQAeYyA3gFDJqYQBcyAzmFKBpcgvjMBvUyyDZUwwMABsIAfj7NW5AL7lyYAJ4AHFAHkQAIQgxcUCACUIIACbQAYgFcQCEfmQBeZAAo1NPAWL10Wb0hiAEoXAD4yZE8A-CCwPzgoETgxAB5-HFiiMAjFck4QJmR8PQMjUwtoHThzem1SwxMzSyhbe0cQF3dQ5wiKKiMwGyhO-qooryzfMnZx5BFxOmQAIgBbFQAVUQllgBolOaoAeiPkDYALYAZkAHdDAGtrgCMIBDgbBhRcGGQQfABafjABwgCAMa6gZCDYadVQaW7nMzIODgnggOSHZAnZHIGB2BzAJxwlBgXDIv5gRFQXH4jrzdQQWbjJ5iOD0ZYs3BPZZMxRURR5ApFEr6RoVFoAcVwuFqyHqovKzWstMJnVcbmC9FIExiPni1CSwBS6UmerC8nCMwGECGIyt42imT1WqZwm2SzWm3dewOh2xADlcDd5pdriwMOcwHixGIVLi4MAxAxdiG4CBHvMyZSUEQ1GJgaIoTaYfSNK7kCy2StOdzefs5k8agBBBgAdQgMfo8CTjP5CiU+Xwwt0CqalSg1XMACYum4NZaxtC7WMqI7AtkXZiFhJ2estosfeXsRcrrcHs9Xu9PsUfn8QIDZCCwRDOkvYQyEUiUQw0Ri-ac4BpdpVVLEkyTTXBs2pPFgKJBly0rdkax5OZ5HrZlmzbDsxHoZgbF7ZABVCBg4BEBhuDBOURzKMcWjaAl8CAA).

Reply:
> I'm also seeing that satisfies isn't strict as well.

Interesting. It seems to be another symptom of the same problem. More on this here:
* https://stackoverflow.com/questions/63006497/typescript-why-not-strict-return-type
* https://stackoverflow.com/questions/57214325/typescript-excess-properties-are-not-checked-in-lambdas-without-explicit-return

> I'm thinking: maybe we can remove the union by setting the onBeforeRender's pageContext to PageContextServer

That sounds good indeed. I'll give that a try.

Reply:
> If the user always extends `PageContext{Client,Server}` by using `Vike.PageContext` then I'm thinking we can even remove the generic type.

And now this sentence just clicked. I didn't fully understand it before. We're on the same page :+1:

Reply:
👍 Let me know if you have other questions.